source/uwp/Renderer/lib/XamlBuilder.cpp

```diff
@@ -886,47 +888,73 @@ namespace AdaptiveNamespace
+
         HRESULT hr = XamlHelpers::IterateOverVector<IAdaptiveCardElement>(children, ancestorHasFallback, [&](IAdaptiveCardElement* element) {
             HRESULT hr = S_OK;
-            HString elementType;
-            RETURN_IF_FAILED(element->get_ElementTypeString(elementType.GetAddressOf()));
-            ComPtr<IAdaptiveElementRendererRegistration> elementRenderers;
-            RETURN_IF_FAILED(renderContext->get_ElementRenderers(&elementRenderers));
-            ComPtr<IAdaptiveElementRenderer> elementRenderer;
-            RETURN_IF_FAILED(elementRenderers->Get(elementType.Get(), &elementRenderer));
+
+            // Get fallback state
             ABI::AdaptiveNamespace::FallbackType elementFallback;
             element->get_FallbackType(&elementFallback);
             const bool elementHasFallback = (elementFallback != FallbackType_None);
             renderArgs->put_AncestorHasFallback(elementHasFallback || ancestorHasFallback);
-            if (elementRenderer != nullptr)
+
+            // Check to see if element's requirements are being met
+            boolean requirementsMet;
+            RETURN_IF_FAILED(element->MeetsRequirements(featureRegistration.Get(), &requirementsMet));
+            hr = requirementsMet ? S_OK : E_PERFORM_FALLBACK;
+
+            // Get element renderer
+            ComPtr<IAdaptiveElementRendererRegistration> elementRenderers;
+            RETURN_IF_FAILED(renderContext->get_ElementRenderers(&elementRenderers));
+            ComPtr<IAdaptiveElementRenderer> elementRenderer;
+            HString elementType;
+            RETURN_IF_FAILED(element->get_ElementTypeString(elementType.GetAddressOf()));
+            RETURN_IF_FAILED(elementRenderers->Get(elementType.Get(), &elementRenderer));
+
+            ComPtr<IAdaptiveHostConfig> hostConfig;
+            RETURN_IF_FAILED(renderContext->get_HostConfig(&hostConfig));
+
+            if (SUCCEEDED(hr) && elementRenderer != nullptr)
             {
-                ComPtr<IAdaptiveHostConfig> hostConfig;
-                RETURN_IF_FAILED(renderContext->get_HostConfig(&hostConfig));
-                // First element does not need a separator added
-                if (currentElement++ > 0)
-                {
-                    bool needsSeparator;
-                    UINT spacing;
-                    UINT separatorThickness;
-                    ABI::Windows::UI::Color separatorColor;
-                    GetSeparationConfigForElement(element, hostConfig.Get(), &spacing, &separatorThickness, &separatorColor, &needsSeparator);
-                    if (needsSeparator)
-                    {
-                        auto separator = CreateSeparator(renderContext, spacing, separatorThickness, separatorColor);
-                        XamlHelpers::AppendXamlElementToPanel(separator.Get(), parentPanel);
-                    }
-                }
+                AddSeparatorIfNeeded(iElement, element, hostConfig, renderContext, parentPanel);
```

Comment:
>AddSeparatorIfNeeded [](start = 16, length = 20)

This throws.... we already have a bug for fixing up HRESULTs vs throwing, right?

Reply:
sure do -- #2534 :)

---
In reply to: [275451172](https://github.com/Microsoft/AdaptiveCards/pull/2684#discussion_r275451172) [](ancestors = 275451172)